client/asset/btc/spv.go

```diff
@@ -0,0 +1,1601 @@
+	acctNum     uint32
+	neutrinoDB  walletdb.DB
+
+	txBlocksMtx sync.Mutex
+	txBlocks    map[chainhash.Hash]*hashEntry
+
+	checkpointMtx sync.Mutex
+	checkpoints   map[outPoint]*scanCheckpoint
+
+	log    dex.Logger
+	loader *wallet.Loader
+}
+
+var _ Wallet = (*spvWallet)(nil)
+
+// loadSPVWallet loads an existing wallet.
+func loadSPVWallet(dbDir string, logger dex.Logger, connectPeers []string, chainParams *chaincfg.Params) (*spvWallet, error) {
+	w, loader, chainService, neutrinoDB, err := loadChainClient(&spvConfig{
+		dbDir:        dbDir,
+		chainParams:  chainParams,
+		connectPeers: connectPeers,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &spvWallet{
+		chainParams: chainParams,
+		cl:          chainService,
+		chainClient: chain.NewNeutrinoClient(chainParams, chainService),
+		acctNum:     0,
+		wallet:      &walletExtender{w, chainParams},
+		neutrinoDB:  neutrinoDB,
+		txBlocks:    make(map[chainhash.Hash]*hashEntry),
+		checkpoints: make(map[outPoint]*scanCheckpoint),
+		log:         logger,
+		loader:      loader,
+	}, nil
+}
+
+// storeTxBlock stores the block hash for the tx in the cache.
+func (w *spvWallet) storeTxBlock(txHash, blockHash chainhash.Hash) {
+	w.txBlocksMtx.Lock()
+	defer w.txBlocksMtx.Unlock()
+	w.txBlocks[txHash] = &hashEntry{
+		hash:       blockHash,
+		lastAccess: time.Now(),
+	}
+}
+
+// txBlock attempts to retrieve the block hash for the tx from the cache.
+func (w *spvWallet) txBlock(txHash chainhash.Hash) (chainhash.Hash, bool) {
+	w.txBlocksMtx.Lock()
+	defer w.txBlocksMtx.Unlock()
+	entry, found := w.txBlocks[txHash]
+	if !found {
+		return chainhash.Hash{}, false
+	}
+	entry.lastAccess = time.Now()
+	return entry.hash, found
+}
+
+// cacheCheckpoint caches a *filterScanResult so that future scans can be
+// skipped or shortened.
+func (w *spvWallet) cacheCheckpoint(txHash *chainhash.Hash, vout uint32, res *filterScanResult) {
+	if res.spend != nil && res.blockHash == nil {
+		// Probably set the start time too late. Don't cache anything
+		return
+	}
+	w.checkpointMtx.Lock()
+	defer w.checkpointMtx.Unlock()
+	w.checkpoints[newOutPoint(txHash, vout)] = &scanCheckpoint{
+		res:        res,
+		lastAccess: time.Now(),
+	}
+}
+
+// checkpoint returns any cached *filterScanResult for the outpoint.
+func (w *spvWallet) checkpoint(txHash *chainhash.Hash, vout uint32) *filterScanResult {
+	w.checkpointMtx.Lock()
+	defer w.checkpointMtx.Unlock()
+	check, found := w.checkpoints[newOutPoint(txHash, vout)]
+	if !found {
+		return nil
+	}
+	check.lastAccess = time.Now()
+	return check.res
+}
+
+// checkpointBlock returns a filterScanResult and the checkpoint block hash. If
+// a result is found with an orphaned checkpoint block hash, it is cleared from
+// the cache and not returned.
+func (w *spvWallet) checkpointBlock(txHash *chainhash.Hash, vout uint32) (*filterScanResult, *chainhash.Hash) {
+	res := w.checkpoint(txHash, vout)
+	if res == nil {
+		return nil, nil
+	}
+	blockHash := &res.checkpoint
```

Comment:
It's not clear why you there is the second Hash output since the `filterScanResult` contains the hash.
There's only one caller of `checkpointBlock`, and it requires the whole `filterScanResult`.

Reply:
Oh. I was originally returning the nearest mainchain ancestor block hash, but then I realized that I would need to clear `filterScanResult` fields if there were orphaned blocks containing outputs or spends, and that looked like it was going to be messy, so I decided just to invalidate the entire checkpoint instead.